### ISSUE TITLE
[Ubuntu 16.04 ARM64] /var/lib/lxd/unix.socket has no 'w' perms for group.

### ISSUE THREAD

**Issue by manjo-git** [lxc/lxd#1551]
On latest Xenial (16.04) /var/lib/lxd/unix.socket does not have 'w' perms for group. This causes lxc command and lxd-images import  commands to fail. I am running this on an ARM64 system have not checked x64 or other archs. 

$ ls -lh /var/lib/lxd/unix.socket
srw-r----- 1 root lxd 0 Nov 18 07:59 /var/lib/lxd/unix.socket
$ lxc list
error: Get http://unix.socket/1.0/containers?recursion=1: dial unix /var/lib/lxd/unix.socket: connect: permission denied
$ sudo chmod ug+rw /var/lib/lxd/unix.socket
$ ls -lh /var/lib/lxd/unix.socket
srw-rw---- 1 root lxd 0 Nov 18 07:59 /var/lib/lxd/unix.socket
$ lxc list
+------+-------+------+------+-----------+-----------+
| NAME | STATE | IPV4 | IPV6 | EPHEMERAL | SNAPSHOTS |
+------+-------+------+------+-----------+-----------+

**Comment by stgraber** [comment]
The LXD code itself clearly sets the socket 0660 and so does the systemd unit, so I have no idea what changed that permission for you...

**Comment by manjo-git** [comment]
This is weird ... this is something else I notice, when I stop lxd does not remove unix.socket, if you remove it manually it fails to restart because looks like unix.socket was not created by lxd when it started. Also if I manually add g+w to unix.socket on system reboot the perms are reset to 640 for unix.socket.

$ apt-cache policy lxd
lxd:
  Installed: 2.0.0~beta1-0ubuntu3

** Stop LXD** 
$ sudo systemctl stop lxd                                       
Warning: Stopping lxd.service, but it can still be activated by:
  lxd-unix.socket

** Status reports stopped **
$ sudo systemctl status lxd                                     
��● lxd.service - Container hypervisor based on LXC
   Loaded: loaded (/lib/systemd/system/lxd.service; indirect; vendor preset: en)
   Active: inactive (dead) since Fri 2016-01-29 16:37:10 UTC; 6s ago
  Process: 2131 ExecStartPost=/usr/bin/lxd waitready --timeout=600 (code=exited)
  Process: 2130 ExecStart=/usr/bin/lxd --group lxd --logfile=/var/log/lxd/lxd.l)
 Main PID: 2130 (code=exited, status=0/SUCCESS)

Jan 29 16:37:10 bandera systemd[1]: lxd.service: Installed new job lxd.serv...59
Jan 29 16:37:10 bandera systemd[1]: lxd.service: Enqueued job lxd.service/s...59
Jan 29 16:37:10 bandera systemd[1]: lxd.service: Changed running -> stop-sigterm
Jan 29 16:37:10 bandera systemd[1]: Stopping Container hypervisor based on .....
Jan 29 16:37:10 bandera systemd[1]: lxd.service: Child 2130 belongs to lxd....ce
Jan 29 16:37:10 bandera systemd[1]: lxd.service: Main process exited, code=...SS
Jan 29 16:37:10 bandera systemd[1]: lxd.service: Changed stop-sigterm -> dead
Jan 29 16:37:10 bandera systemd[1]: lxd.service: Job lxd.service/stop finis...ne
Jan 29 16:37:10 bandera systemd[1]: Stopped Container hypervisor based on LXC.
Jan 29 16:37:10 bandera systemd[1]: lxd.service: cgroup is empty

** Was expecting /unix.socket would be removed **
$ ls -lh /var/lib/lxd/unix.socket 
srw-r----- 1 root lxd 0 Nov 18 07:59 /var/lib/lxd/unix.socket

** manually removing unix.socket **
$ sudo mv /var/lib/lxd/unix.socket /var/lib/lxd/old.unix.socket 

** Start LXD Failed **
$ sudo systemctl start lxd                                      
Job for lxd.service failed because a timeout was exceeded. See "systemctl statu.
$ sudo systemctl status lxd
��● lxd.service - Container hypervisor based on LXC
   Loaded: loaded (/lib/systemd/system/lxd.service; indirect; vendor preset: en)
   Active: activating (start-post) since Fri 2016-01-29 16:42:24 UTC; 1min 22s o
 Main PID: 2335 (lxd);         : 2336 (lxd)
    Tasks: 0 (limit: 512)
   Memory: 1.0M
      CPU: 39ms
   CGroup: /system.slice/lxd.service
           ��‣ 2335 /usr/bin/lxd --group lxd --logfile=/var/log/lxd/lxd.log
           ��‣ 2336 /usr/bin/lxd waitready --timeout=600

Jan 29 16:42:24 bandera systemd[1]: lxd.service: Converting job lxd.service...rt
Jan 29 16:42:24 bandera systemd[1]: lxd.service: About to execute: /usr/bin...og
Jan 29 16:42:24 bandera systemd[1]: lxd.service: Forked /usr/bin/lxd as 2335
Jan 29 16:42:24 bandera systemd[1]: lxd.service: About to execute: /usr/bin...00
Jan 29 16:42:24 bandera systemd[1]: lxd.service: Forked /usr/bin/lxd as 2336
Jan 29 16:42:24 bandera systemd[1]: lxd.service: Changed dead -> start-post
Jan 29 16:42:24 bandera systemd[1]: Starting Container hypervisor based on .....
Jan 29 16:42:24 bandera systemd[2335]: lxd.service: Executing: /usr/bin/lxd...og
Jan 29 16:42:24 bandera systemd[2336]: lxd.service: Executing: /usr/bin/lxd...00
Jan 29 16:42:24 bandera lxd[2335]: t=2016-01-29T16:42:24+0000 lvl=warn msg=...."
Hint: Some lines were ellipsized, use -l to show in full.
$ ls -lh /var/lib/lxd/unix.socket                 
ls: cannot access /var/lib/lxd/unix.socket: No such file or directory
$ lxc list
Generating a client certificate. This may take a minute...
If this is your first run, you will need to import images using the 'lxd-images.
For example: 'lxd-images import ubuntu --alias ubuntu'.
error: Get http://unix.socket/1.0/containers?recursion=1: dial unix /var/lib/lxy

** Manually add unix.socket **
$ sudo mv /var/lib/lxd/old.unix.socket /var/lib/lxd/unix.socket

** LXD starts again **
$ sudo systemctl status lxd
● lxd.service - Container hypervisor based on LXC
   Loaded: loaded (/lib/systemd/system/lxd.service; indirect; vendor preset: enabled)
   Active: active (running) since Fri 2016-01-29 16:54:28 UTC; 41s ago
  Process: 2615 ExecStartPost=/usr/bin/lxd waitready --timeout=600 (code=exited, status=0/SUCCESS)
 Main PID: 2614 (lxd)
    Tasks: 0 (limit: 512)
   Memory: 680.0K
      CPU: 37ms
   CGroup: /system.slice/lxd.service
           ‣ 2614 /usr/bin/lxd --group lxd --logfile=/var/log/lxd/lxd.log

Jan 29 16:54:28 bandera systemd[2615]: lxd.service: Executing: /usr/bin/lxd...00
Jan 29 16:54:28 bandera lxd[2614]: t=2016-01-29T16:54:28+0000 lvl=warn msg=...."
Jan 29 16:54:28 bandera systemd[1]: lxd.service: Child 2615 belongs to lxd....ce
Jan 29 16:54:28 bandera systemd[1]: lxd.service: Control process exited, co...=0
Jan 29 16:54:28 bandera systemd[1]: lxd.service: Got final SIGCHLD for stat...t.
Jan 29 16:54:28 bandera systemd[1]: lxd.service: Changed start-post -> running
Jan 29 16:54:28 bandera systemd[1]: lxd.service: Job lxd.service/start fini...ne
Jan 29 16:54:28 bandera systemd[1]: Started Container hypervisor based on LXC.
Jan 29 16:54:28 bandera systemd[1]: lxd.service: cgroup is empty
Jan 29 16:54:28 bandera systemd[1]: lxd.service: Failed to send unit change...er
Hint: Some lines were ellipsized, use -l to show in full.

** lxc list fails because group has no write perms **
$ lxc list
error: Get http://unix.socket/1.0/containers?recursion=1: dial unix /var/lib/lxd/unix.socket: connect: permission denied

** manually add write perms and rerun lxc list **
$ sudo chmod g+w /var/lib/lxd/unix.socket
$ lxc list
+------+-------+------+------+-----------+-----------+
| NAME | STATE | IPV4 | IPV6 | EPHEMERAL | SNAPSHOTS |
+------+-------+------+------+-----------+-----------+

**Comment by manjo-git** [comment]
Wonder what is messing with unix.socket because clearly systemd seems to have the right bits set.
/etc/systemd/system/sockets.target.wants$ cat lxd-unix.socket 
[Socket]
ListenStream=/var/lib/lxd/unix.socket
SocketGroup=lxd
SocketMode=0660
Service=lxd.service

[Install]
WantedBy=sockets.target

**Comment by stgraber** [comment]
lxd is socket activated. That means that by default the lxd unit won't be running but the lxd-unix.socket one will, that second unit is the one creating and binding /var/lib/lxd/unix.socket.
On connection to that socket, the main lxd unit is spawned and the pre-initialized fd is handed over to it.

When you stop the lxd unit, the daemon is stopped, but the socket is still open and listening through the lxd-unix.socket unit. That's why it's not removed.

If you remove the socket, then lxd-unix.socket is still running but without the filesystem entry which causes lxd to fail to start.

In short, don't manually mess with /var/lib/lxd/unix.socket when you're using the system units :)

If you want to manually start lxd, you need to stop both the lxd and lxd-unix.socket units, then you can manually start lxd.

**Comment by stgraber** [comment]
hey @manjo-git any update on this?

**Comment by manjo-git** [comment]
I was unable to determine what messes with the group permission for /var/lib/lxd/unix.socket so I worked around it in arm64 xenial in /etc/systemd/system/multi-user.target.wants/lxd-startup.service by adding 
``
ExecStart=/bin/chmod g+w /var/lib/lxd/unix.socket
ExecStart=/usr/bin/lxd activateifneeded
``
Because if I set group 'w' permissions and I reboot, something resets group perms to 'r' only. 

Like you said in the previous post, systemd and lxd seems to do the right things, so I am not sure what else could mess up the group permissions on the unix.socket.

**Comment by stgraber** [comment]
I'm going to close this LXD issue as we're clearly doing the right thing everywhere I can see, my best guess is that something's wrong in systemd or something's changing those permissions behind both systemd and lxd's backs...

**Comment by manjo-git** [comment]
Yes I agree. It can be closed for now.
